bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/AbstractRetryingOperation.java

```diff
@@ -329,16 +329,36 @@ protected void run() {
     return call;
   }
 
-  protected CallOptions getCallOptions() {
-    if (callOptions.getDeadline() != null) {
-      return callOptions;
-    }
-    if (request instanceof ReadRowsRequest
-        && !CallOptionsFactory.ConfiguredCallOptionsFactory.isGet((ReadRowsRequest) request)) {
-      // This is a streaming read.
-      return callOptions;
+  /**
+   * Returns the {@link CallOptions} that a user set for the entire Operation, which can span multiple RPCs/retries.
+   * @return The {@link CallOptions}
+   */
+  protected CallOptions getOperationCallOptions() {
+    return callOptions;
+  }
+
+  /**
+   * Create an {@link CallOption} that has a fail safe RPC deadline to make sure that unary operations don't hang. This
+   * will have to be overridden for streaming RPCs like read rows.
+   *
+   * @return a {@link CallOptions}
+   */
+  protected CallOptions getRpcCallOptions() {
```

Comment:
I am confused. The logic here stays the same, no?

Can you document how user specified deadline is used; streaming vs non-streaming?

For each RPC:

```
               | User deadline provided            |  User deadline not provided
-----------------------------------------------------------------------------
Streaming      |    User deadline                  |  watch dog?
-----------------------------------------------------------------------------
Non-Streaming  |    User deadline                  |  UNARY_DEADLINE_MINUTES
-----------------------------------------------------------------------------

```